subsys/storage/flash_map/flash_map.c

```diff
@@ -0,0 +1,270 @@
+{
+	for (int i = 0; i < ARRAY_SIZE(flash_drivers_map); i++) {
+		if (flash_drivers_map[i].id == id) {
+			return device_get_binding(flash_drivers_map[i].name);
+		}
+	}
+
+	k_panic();
+}
+
+static inline bool is_in_flash_area_bounds(const struct flash_area *fa,
+					   off_t off, size_t len)
+{
+	return (off <= fa->fa_size && off + len <= fa->fa_size);
+}
+
+struct flash_area const *get_flash_area_from_id(int idx)
+{
+	for (int i = 0; i < flash_map_entries; i++) {
+		if (flash_map[i].fa_id == idx) {
+			return &flash_map[i];
+		}
+	}
+
+	return NULL;
+}
+
+#if defined(CONFIG_FLASH_PAGE_LAYOUT)
+/*
+ * Check if a flash_page_foreach() callback should exit early, due to
+ * one of the following conditions:
+ *
+ * - The flash page described by "info" is before the area of interest
+ *   described in "data"
+ * - The flash page is after the end of the area
+ * - There are too many flash pages on the device to fit in the array
+ *   held in data->ret. In this case, data->status is set to -ENOMEM.
+ *
+ * The value to return to flash_page_foreach() is stored in
+ * "bail_value" if the callback should exit early.
+ */
+static bool should_bail(const struct flash_pages_info *info,
+						struct layout_data *data,
+						bool *bail_value)
+{
+	if (info->start_offset < data->area_off) {
+		*bail_value = true;
+		return true;
+	} else if (info->start_offset >= data->area_off + data->area_len) {
+		*bail_value = false;
+		return true;
+	} else if (data->ret_idx >= data->ret_len) {
+		data->status = -ENOMEM;
+		*bail_value = false;
+		return true;
+	}
+
+	return false;
+}
+
+/*
+ * Generic page layout discovery routine. This is kept separate to
+ * support both the deprecated flash_area_to_sectors() and the current
+ * flash_area_get_sectors(). A lot of this can be inlined once
+ * flash_area_to_sectors() is removed.
+ */
+static int flash_area_layout(int idx, int *cnt, void *ret,
+flash_page_cb cb, struct layout_data *cb_data)
+{
+	struct device *flash_dev;
+
+	cb_data->area_idx = idx;
+
+	const struct flash_area *fa;
+
+	fa = get_flash_area_from_id(idx);
+	if (fa == NULL) {
+		return -EINVAL;
+	}
+
+	cb_data->area_idx = idx;
+	cb_data->area_off = fa->fa_off;
+	cb_data->area_len = fa->fa_size;
+
+	cb_data->ret = ret;
+	cb_data->ret_idx = 0;
+	cb_data->ret_len = *cnt;
+	cb_data->status = 0;
+
+	flash_dev = get_flah_dev_from_id(fa->fa_device_id);
+
+	flash_page_foreach(flash_dev, cb, cb_data);
+
+	if (cb_data->status == 0) {
+		*cnt = cb_data->ret_idx;
+	}
+
+	return cb_data->status;
+}
+
+static bool get_sectors_cb(const struct flash_pages_info *info, void *datav)
+{
+	struct layout_data *data = datav;
+	struct flash_sector *ret = data->ret;
+	bool bail;
+
+	if (should_bail(info, data, &bail)) {
+		return bail;
+	}
+
+	ret[data->ret_idx].fs_off = info->start_offset - data->area_off;
+	ret[data->ret_idx].fs_size = info->size;
+	data->ret_idx++;
+
+	return true;
+}
+
+int flash_area_get_sectors(int idx, uint32_t *cnt, struct flash_sector *ret)
```

Comment:
@mbolivar Should I add Linaro to copy right list? I mean that I take this form mcuboot, and it was your code.

Reply:
I think the main thing to do is preserve any copyright notices that were in the files which you're copying content from, so if there's Linaro headers in the original file, please make sure they're reproduced here, yes.

Reply:
So I will keep this as is - As mcuboot file from which i take this code doesn't have copyright list.